tests/ui/redundant_locals.rs

```diff
@@ -0,0 +1,97 @@
+#![allow(unused, clippy::no_effect)]
+#![warn(clippy::redundant_locals)]
+
+fn main() {}
+
+fn immutable() {
+    let x = 1;
+    let x = x;
+}
+
+fn mutable() {
+    let mut x = 1;
+    let mut x = x;
+}
+
+fn upgraded_mutability() {
+    let x = 1;
+    let mut x = x;
+}
+
+fn downgraded_mutability() {
+    let mut x = 1;
+    let x = x;
+}
+
+fn coercion(par: &mut i32) {
+    let par: &i32 = par;
+
+    let x: &mut i32 = &mut 1;
+    let x: &i32 = x;
+}
+
+fn parameter(x: i32) {
+    let x = x;
+}
+
+fn many() {
+    let x = 1;
+    let x = x;
+    let x = x;
+    let x = x;
+    let x = x;
+}
+
+fn interleaved() {
+    let a = 1;
+    let b = 2;
+    let a = a;
+    let b = b;
+}
+
+fn block() {
+    {
+        let x = 1;
+        let x = x;
+    }
+}
+
+fn closure() {
+    || {
+        let x = 1;
+        let x = x;
+    };
+    |x: i32| {
+        let x = x;
+    };
+}
+
+fn consequential_drop_order() {
+    use std::sync::Mutex;
+
+    let mutex = Mutex::new(1);
+    let guard = mutex.lock().unwrap();
+
+    {
+        let guard = guard;
+    }
+}
+
+fn inconsequential_drop_order() {
+    let x = 1;
+
+    {
+        let x = x;
+    }
+}
+
+fn macros() {
```

Comment:
I think tests for external/procedural macros are worthwhile. Add `//@aux-build:proc_macros.rs` to the top of the file, then `extern crate proc_macros`. This will let you use the macros in `tests/ui/auxiliary/proc_macros.rs`.

Something like
```rust
external! {
	let x = 1;
	let x = x;
}
// This will set the token's spans to the end of `span`, thus testing that `is_from_proc_macro` works
with_span! {
	span
	let x = 1;
	let x = x;
}
```